### PR TITLE
hotfix for extensions on nested fields

### PR DESCRIPTION
**Pull request by jeremiepiotte-districtm**
Extensions on nested fields were not handled properly.  This PR revisits the previous extension support implementation and hotfixes extensions on nested fields.

I also added a bigger unit test that will go through all the edge cases.

### CODE REVIEW

File: plugin_test.go [hunk 1]
```diff
@@ -227,6 +227,83 @@ func TestMessageWithExtensionOnNestedField(t *testing.T) {
 		})
 }
 
+// Try a complex message with extension with a few edge cases
+func TestMessagesWithExtensionExceptions(t *testing.T) {
+	testConvert(t, `
```
Comment (QuentinMcgaw-districtm) [inline plugin_test.go:232]: just nitting around, let's keep adding those `t.Parallel()` to detect data races

```suggestion
func TestMessagesWithExtensionExceptions(t *testing.T) {
	t.Parallel()
	testConvert(t, `
```

File: plugin_test.go [hunk 2]
```diff
@@ -227,6 +227,83 @@ func TestMessageWithExtensionOnNestedField(t *testing.T) {
...
+					name: "qux"
+					field: < name: "a" number: 1 label: LABEL_REQUIRED type: TYPE_STRING json_name: "a" >
+				>
+				extension: <
+					name: "baz"
+					number: 126
+					label: LABEL_OPTIONAL
+					type: TYPE_INT32
+					extendee: ".example_package"
+					json_name: "baz"
```
Comment (QuentinMcgaw-districtm) [inline plugin_test.go:289]: (Sorry for my ignorant question) Why is the baz extension not in the result?

**Review by QuentinMcgaw-districtm — APPROVED** [review]

File: main.go [hunk 3]
```diff
@@ -484,21 +484,21 @@ func addExtensions(msg *descriptor.DescriptorProto, extensions []*descriptor.Fie
...
+			nestedMsgs := msg.GetNestedType()
+			foundFlag = false
 			for _, nestedMsg := range nestedMsgs {
 				if nestedMsg.GetName() == extendeeNodes[i] {
-					foundNode = nestedMsg
+					msg = nestedMsg
+					foundFlag = true
 					break
 				}
 			}
```
Comment (EricSeguin-districtm) [inline main.go:498]: Could be wrong but I think the outer loop here is missing a break statement:

```suggestion
			}

			if foundFlag {
				break
			}
```

File: plugin_test.go (hunk 2 above)
Comment (jeremiepiotte-districtm) [reply plugin_test.go:290]: because quz.proto is not in `file_to_generate `

File: main.go (hunk 3 above)
Comment (jeremiepiotte-districtm) [reply main.go:498]: I see you are paying attention!  But no, this is wanted behavior.  We don't want to break the outer-loop because we want to keep going as deep as possible.  It's not because we found a match that it's the last possible one.

Comment (EricSeguin-districtm) [reply main.go:498]: Right, I missed how `msg` is progressively set to deeper and deeper nested levels. However, the flag behaviour still seems bugged since it gets reset on each iteration of the outer loop. What happens if there's a match at `i == 1` but none at `i == 2`? If I'm reading the code right, `foundFlag` will stay false and the matched node will not be set on line 102.

Comment (jeremiepiotte-districtm) [reply main.go:498]: Yep, you are right, that's what happens. Thing is, that's also wanted behavior due to context.  We only want to consider a match if everything matches until the end of the outer-loop

Comment (EricSeguin-districtm) [reply main.go:498]: Alright! It's unfortunate that you merged this PR already, I would have asked you to add a note documenting that this is the desired behaviour. Without that context, it's hard to understand what the desired behaviour is.

Comment (jeremiepiotte-districtm) [reply main.go:498]: One could argue that tested code is the desired behaviour. There are tests that test specifically this behaviour.  But I do get your point, sorry if I merged too soon, I thought everyone was happy with it and honestly really wanted to get past this

Comment (EricSeguin-districtm) [reply main.go:498]: Sorry if that came across as accusatory, that wasn't the intention! FWIW I disagree with tests serving as documentation, they can definitely be a reference but the implementation should ideally be clear on its own or documented if it can't be. Just wanted to outline that here but it's not a big deal in this case :)
